dashboard/lib/user_multi_auth_helper.rb

```diff
@@ -24,6 +24,23 @@ def oauth_tokens_for_provider(provider)
     end
   end
 
+  def update_oauth_credential_tokens(auth_hash)
+    # No-op if auth_hash does not contain credentials
+    return unless auth_hash.key?(:credentials)
+
+    credentials_hash = auth_hash[:credentials]
+    if migrated?
+      auth_option = authentication_options.find_by(credential_type: auth_hash[:provider], user_id: id)
```

Comment:
I think this line is problematic because it assumes that the user has _at most_ 1 AuthenticationOption per provider. We're using this same logic [here](https://github.com/code-dot-org/code-dot-org/blob/853754f55e72073c921d7522d868f8a5465f0527/dashboard/lib/user_multi_auth_helper.rb#L8-L11) because in this context, we don't have access to search on provider+uid like we normally would. 

I'm not really sure how to solve this or what we want to do about it, but wanted to call it out for feedback.

Reply:
That seems like a thing we should fix pre-merge or it could be destructive.

We've got the relevant uid in the sign-in case, so we could update the appropriate authoption there.  I take it we don't have the relevant uid in the classroom sync case precisely because classroom sync is currently assuming you only have one google account linked, so there's sort of an upstream problem there.  I'm not sure what the solution is (UI to ask the user which account to use?) but I'd suggest the following action plan:

1. For now, only update the tokens on sign-in, not on classroom sync.
2. Let's go fix classroom sync assuming only one authoption.
3. We can come back and update tokens on classroom sync after that.

But I'm not clear in the first place on why we'd want to set tokens on classroom sync, so maybe this is a bad plan.

Reply:
I think that's a good plan given the situation. I'm not _sure_ why the tokens are updated after classroom sync, but I would assume it's because we don't log users out of Code.org, so being able to refresh tokens without them having to log out and back in is beneficial.

@joshlory, do you remember why we update user OAuth tokens [here](https://github.com/code-dot-org/code-dot-org/blob/04c162b0fa316ca798f883de681fff72b01e9e42/dashboard/app/controllers/api_controller.rb#L68-L73) after querying google classroom? this code is over a year old, so it's understandable if you don't 😂

Reply:
I believe the token can change when the user authorizes additional scopes.  But not 100% sure...

Reply:
This might have a bit more info? https://github.com/code-dot-org/code-dot-org/commit/1170e045650720bdc61f5ada29bdd0ef2d2c401a#diff-d0ba960dee0b3ab3f04aebc9fb0e01c6R30

Reply:
Interesting, I'll look into that to confirm. If that is the case, we won't want to disable updating tokens even temporarily, so good to know. Thanks, Josh!

Reply:
That commit helps, too. I wasn't sure if we _needed_ to update the token or if it was a convenience thing (more like, "we update the tokens because we can and have received fresher tokens")